File: backend/app/services/file_parser.py
import csv
import io
from typing import Any

COLUMN_ALIASES = {
    "name": ["name", "productname", "product", "product_name"],
    "sku": ["sku", "productcode", "product_code"],
    "category": ["category", "productcategory", "product_category"],
    "description": ["description"],
    "technical_specs": ["technicalspecs", "technical_specs", "specifications", "specs"],
    "certifications": ["certifications", "certs"],
    "moq": ["moq", "minimumorderquantity", "minimum_order_quantity", "minqty"],
    "pricing": ["pricing"],
    "lead_time_days": ["leadtime", "lead_time", "leadtime_days", "lead_time_days", "deliverydays"],
}


def _normalize(key: str) -> str:
    return key.lower().replace("_", "").replace("-", "").replace(" ", "")


def _parse_number(val: str) -> int | None:
    if not val or not val.strip():
        return None
    cleaned = "".join(c for c in val if c.isdigit() or c in ".-")
    if not cleaned:
        return None
    try:
        return int(float(cleaned))
    except ValueError:
        return None


def _resolve_column(headers: list[str]) -> dict[str, int]:
    header_map: dict[str, str] = {}
    for h in headers:
        header_map[_normalize(h)] = h

    mapping: dict[str, int] = {}
    for field, aliases in COLUMN_ALIASES.items():
        for alias in aliases:
            norm = _normalize(alias)
            if norm in header_map:
                mapping[field] = headers.index(header_map[norm])
                break
    return mapping


def _rows_to_products(rows: list[list[str]]) -> list[dict[str, Any]]:
    if len(rows) < 2:
        raise ValueError("File must have a header row and at least one data row")

    headers = rows[0]
    col_map = _resolve_column(headers)

    if "name" not in col_map:
        raise ValueError("File must have a 'name' column")

    products: list[dict[str, Any]] = []
    for row in rows[1:]:
        if not row or all(not cell.strip() for cell in row):
            continue

        def get_val(field: str) -> str:
            idx = col_map.get(field)
            return row[idx].strip() if idx is not None and idx < len(row) else ""

        name = get_val("name")
        if not name:
            continue

        category = get_val("category") or "other"
        category = category.lower().replace(" ", "_")

        products.append({
            "name": name,
            "sku": get_val("sku") or None,
            "category": category,
            "description": get_val("description") or None,
            "technical_specs": get_val("technical_specs") or None,
            "certifications": get_val("certifications") or None,
            "moq": _parse_number(get_val("moq")),
            "unit_price": None,
            "price_range_low": None,
            "price_range_high": None,
            "pricing": get_val("pricing") or None,
            "lead_time_days": _parse_number(get_val("lead_time_days")),
        })

    if not products:
        raise ValueError("No valid product rows found")

    return products


async def _parse_csv(content: bytes) -> list[dict[str, Any]]:
    text = content.decode("utf-8-sig")
    reader = csv.reader(io.StringIO(text))
    rows = [[cell.strip() for cell in row] for row in reader]
    return _rows_to_products(rows)


async def _parse_excel(content: bytes) -> list[dict[str, Any]]:
    from openpyxl import load_workbook

    wb = load_workbook(io.BytesIO(content), read_only=True, data_only=True)
    ws = wb.active
    rows: list[list[str]] = []
    for row in ws.iter_rows(values_only=True):
        if row is None:
            continue
        rows.append(["" if cell is None else str(cell).strip() for cell in row])
    return _rows_to_products(rows)


async def _parse_docx(content: bytes) -> list[dict[str, Any]]:
    from docx import Document

    doc = Document(io.BytesIO(content))
    rows: list[list[str]] = []
    for table in doc.tables:
        for row in table.rows:
            rows.append([cell.text.strip() for cell in row.cells])
    if not rows:
        raise ValueError("No tables found in Word document; expected a product table")
    return _rows_to_products(rows)


async def _parse_pdf(content: bytes) -> list[dict[str, Any]]:
    from PyPDF2 import PdfReader

    reader = PdfReader(io.BytesIO(content))
    parts: list[str] = []
    for page in reader.pages:
        text = page.extract_text() or ""
        parts.append(text)

    full_text = "\n".join(parts).strip()
    if not full_text:
        raise ValueError("No extractable text found in PDF")

    lines = [line.strip() for line in full_text.splitlines() if line.strip()]
    name = lines[0] if lines else "Imported from PDF"

    return [{
        "name": name,
        "sku": None,
        "category": "other",
        "description": full_text,
        "technical_specs": None,
        "certifications": None,
        "moq": None,
        "unit_price": None,
        "price_range_low": None,
        "price_range_high": None,
        "pricing": None,
        "lead_time_days": None,
    }]


async def parse_file(filename: str, file_content: bytes) -> list[dict[str, Any]]:
    ext = filename.rsplit(".", 1)[-1].lower() if "." in filename else ""

    if ext == "csv":
        return await _parse_csv(file_content)
    if ext == "xlsx":
        return await _parse_excel(file_content)
    if ext == "xls":
        raise ValueError("Legacy .xls format is not supported; please upload .xlsx")
    if ext == "docx":
        return await _parse_docx(file_content)
    if ext == "doc":
        raise ValueError("Legacy .doc format is not supported; please upload .docx")
    if ext == "pdf":
        return await _parse_pdf(file_content)

    raise ValueError(f"Unsupported file type: {ext}")

File: frontend/src/lib/ai/file-parser.ts
/**
 * File parser — simulates extracting products from uploaded files.
 * Swap in: pdf.js / xlsx / mammoth.js for real client-side parsing.
 */

import type { Product } from "@/types"

const MOCK_PRODUCTS: Omit<Product, "id" | "created_at" | "is_active">[] = [
  {
    name: "LED Panel Light 60x60cm",
    sku: "LED-PL6060-EU",
    category: "led_lighting",
    description: "High-quality LED panel light, 40W, 600x600mm, suitable for office and commercial spaces.",
    technical_specs: "Power: 40W, Voltage: 220-240V, Luminous Flux: 4000lm, Color Temperature: 4000K/6500K, CRI>80, Size: 595x595mm",
    certifications: "CE, RoHS, EMC",
    moq: 100,
    unit_price: 12.50,
    price_range_low: 10.00,
    price_range_high: 15.00,
    pricing: null,
    lead_time_days: 25,
    image_url: null,
  },
  {
    name: "LED High Bay Light 150W",
    sku: "LED-HB150-EU",
    category: "led_lighting",
    description: "Industrial grade LED high bay light, 150W, ideal for warehouses and factories. IP65 waterproof.",
    technical_specs: "Power: 150W, Voltage: 85-265V, Luminous Flux: 18000lm, Color Temperature: 5000K, Beam Angle: 90°, IP65",
    certifications: "CE, RoHS, IP65",
    moq: 50,
    unit_price: 45.00,
    price_range_low: 38.00,
    price_range_high: 52.00,
    pricing: null,
    lead_time_days: 30,
    image_url: null,
  },
  {
    name: "LED Strip Light 5050 RGB",
    sku: "LED-ST5050-RGB",
    category: "led_lighting",
    description: "Flexible LED strip light, 5050 SMD RGB, 60LEDs/m, with remote control.",
    technical_specs: "LED Type: 5050 SMD, 60 LEDs/m, Voltage: DC12V, Power: 14.4W/m, RGB, Width: 10mm, Length: 5m/roll",
    certifications: "CE, RoHS",
    moq: 200,
    unit_price: 3.80,
    price_range_low: 3.00,
    price_range_high: 4.50,
    pricing: null,
    lead_time_days: 15,
    image_url: null,
  },
  {
    name: "LED Flood Light 100W",
    sku: "LED-FL100-EU",
    category: "led_lighting",
    description: "Outdoor LED floodlight, 100W, IP66 waterproof, suitable for building facades and parking lots.",
    technical_specs: "Power: 100W, Voltage: 220-240V, Luminous Flux: 10000lm, Color Temperature: 6500K, IP66, Die-cast aluminum",
    certifications: "CE, RoHS, IP66, TUV",
    moq: 50,
    unit_price: 28.00,
    price_range_low: 24.00,
    price_range_high: 32.00,
    pricing: null,
    lead_time_days: 20,
    image_url: null,
  },
  {
    name: "LED Tube Light T8 120cm",
    sku: "LED-T8-120-EU",
    category: "led_lighting",
    description: "T8 LED tube light, 18W, 120cm, direct replacement for fluorescent tubes.",
    technical_specs: "Power: 18W, Voltage: 220-240V, Luminous Flux: 1800lm, Color Temperature: 4000K/6500K, Length: 1200mm",
    certifications: "CE, RoHS, EMC",
    moq: 500,
    unit_price: 2.50,
    price_range_low: 2.00,
    price_range_high: 3.00,
    pricing: null,
    lead_time_days: 15,
    image_url: null,
  },
]

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms))
}

function shuffle<T>(arr: T[]): T[] {
  const a = [...arr]
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]]
  }
  return a
}

export interface ParseResult {
  filename: string
  fileType: string
  products: Omit<Product, "id" | "created_at" | "is_active">[]
}

export async function parseFile(file: File): Promise<ParseResult> {
  const ext = file.name.split(".").pop()?.toLowerCase() || ""

  if (ext === "csv") {
    const products = await parseCSV(file)
    return { filename: file.name, fileType: ext, products }
  }

  // Simulate processing time for other formats
  await sleep(1500)

  let products: typeof MOCK_PRODUCTS

  switch (ext) {
    case "pdf":
      products = MOCK_PRODUCTS.slice(0, 3)
      break
    case "xlsx":
    case "xls":
      products = MOCK_PRODUCTS.slice(2, 5)
      break
    case "docx":
    case "doc":
      products = MOCK_PRODUCTS.slice(3)
      break
    default:
      products = shuffle(MOCK_PRODUCTS).slice(0, 3)
  }

  return {
    filename: file.name,
    fileType: ext,
    products,
  }
}

function parseCSVLine(line: string): string[] {
  const result: string[] = []
  let current = ""
  let inQuotes = false

  for (let i = 0; i < line.length; i++) {
    const ch = line[i]
    if (ch === '"') {
      if (inQuotes && i + 1 < line.length && line[i + 1] === '"') {
        current += '"'
        i++
      } else {
        inQuotes = !inQuotes
      }
    } else if (ch === "," && !inQuotes) {
      result.push(current.trim())
      current = ""
    } else {
      current += ch
    }
  }
  result.push(current.trim())
  return result
}

async function parseCSV(file: File): Promise<Omit<Product, "id" | "created_at" | "is_active">[]> {
  const text = await file.text()
  const lines = text.split(/\r?\n/).filter(line => line.trim())

  if (lines.length < 2) {
    throw new Error("CSV file must have a header row and at least one data row")
  }

  const headers = parseCSVLine(lines[0]).map(h => h.toLowerCase().replace(/[""\s]/g, ""))
  const products: Omit<Product, "id" | "created_at" | "is_active">[] = []

  for (let i = 1; i < lines.length; i++) {
    const values = parseCSVLine(lines[i])
    if (values.length === 0 || values.every(v => !v)) continue

    const row: Record<string, string> = {}
    headers.forEach((h, idx) => {
      row[h] = values[idx]?.trim() || ""
    })

    const name = row["name"] || row["productname"] || row["product"] || row["product_name"] || ""
    if (!name) continue

    const pricing = row["pricing"] || null

    products.push({
      name,
      sku: row["sku"] || row["productcode"] || row["product_code"] || null,
      category: (row["category"] || row["productcategory"] || row["product_category"] || "other").toLowerCase().replace(/\s+/g, "_"),
      description: row["description"] || null,
      technical_specs: row["technicalspecs"] || row["technical_specs"] || row["specifications"] || row["specs"] || null,
      certifications: row["certifications"] || row["certs"] || null,
      moq: parseNumber(row["moq"] || row["minimumorderquantity"] || row["minimum_order_quantity"] || row["minqty"]),
      unit_price: null,
      price_range_low: null,
      price_range_high: null,
      pricing,
      lead_time_days: parseNumber(row["leadtime"] || row["lead_time"] || row["leadtime_days"] || row["lead_time_days"] || row["deliverydays"]),
      image_url: null,
    })
  }

  if (products.length === 0) {
    throw new Error("No valid product rows found in CSV")
  }

  return products
}

function parseNumber(val: string | undefined): number | null {
  if (!val) return null
  const cleaned = val.replace(/[^0-9.\-]/g, "")
  const num = parseFloat(cleaned)
  return isNaN(num) ? null : num
}
